geth/node/manager.go

```diff
@@ -31,86 +34,90 @@ var (
 	ErrRPCClient                   = errors.New("failed to init RPC client")
 )
 
+const (
+	started = 0
+	stopped = 1
+	pending = 2
+)
+
 // NodeManager manages Status node (which abstracts contained geth node)
+// uses interfaces to prevent races on exported objects.
 type NodeManager struct {
-	sync.RWMutex
-	config         *params.NodeConfig // Status node configuration
-	node           *node.Node         // reference to Geth P2P stack/node
-	nodeStarted    chan struct{}      // channel to wait for start up notifications
-	nodeStopped    chan struct{}      // channel to wait for termination notifications
-	whisperService *whisper.Whisper   // reference to Whisper service
-	lesService     *les.LightEthereum // reference to LES service
-	rpcClient      *rpc.Client        // reference to RPC client
+	state *int32
+
+	config atomic.Value // Status node configuration
+
+	node     geth.Node // reference to Geth P2P stack/node
+	nodeLock sync.RWMutex
+
+	nodeStarted     chan struct{} // channel to wait for start up notifications
+	nodeStartedLock sync.RWMutex
+
+	nodeStopped     chan struct{} // channel to wait for termination notifications
+	nodeStoppedLock sync.RWMutex
+
+	whisperService     *whisperv5.Whisper // reference to Whisper service
+	whisperServiceLock sync.RWMutex
+
+	lesService     services.LesService // reference to LES service
+	lesServiceLock sync.RWMutex
+
+	statusBackend services.StatusBackend // reference to ethereum backend
+
+	rpcClient     geth.RPCClient // reference to RPC client
+	rpcClientLock sync.RWMutex
 }
 
 // NewNodeManager makes new instance of node manager
 func NewNodeManager() *NodeManager {
-	m := &NodeManager{}
+	var isStarted int32 = pending
+	m := &NodeManager{state: &isStarted}
+
 	go HaltOnInterruptSignal(m) // allow interrupting running nodes
 
 	return m
 }
 
-// StartNode start Status node, fails if node is already started
+// StartNode start Status node, fails if node is already started.
 func (m *NodeManager) StartNode(config *params.NodeConfig) (<-chan struct{}, error) {
-	m.Lock()
-	defer m.Unlock()
+	if m.isNodeStarted() {
+		return nil, ErrNodeExists
+	}
+
+	m.initLog(config)
 
 	return m.startNode(config)
 }
 
-// startNode start Status node, fails if node is already started
-func (m *NodeManager) startNode(config *params.NodeConfig) (<-chan struct{}, error) {
-	if m.node != nil || m.nodeStarted != nil {
-		return nil, ErrNodeExists
+// StartNodeWait the same as StartNode, but works in sync mode.
+func (m *NodeManager) StartNodeWait(config *params.NodeConfig) error {
```

Comment:
I am not a fan of this kind of methods... It's basically a duplication of `StartNode(config *params.NodeConfig) (<-chan struct{}, error)`.

Reply:
If we do refactoring we shouldn`t change current interfaces and behaviour. Nowedays StartNode, StopNode, RestartNode and ResetChainData works in async mode but we often use them in code like:
```
ch, err := nodeManager.StartNode(...)
<- ch
```

We make an action and wait its complition. So i think we need both Sync and Async methods and then decide where what solution is better.

Reply:
> If we do refactoring we shouldn`t change current interfaces and behaviour

Why not? That was my intention to get rid of the old design and replace it with a new one. Unless this PR is only about fixing the race conditions and locks. 

> So i think we need both Sync and Async methods and then decide where what solution is better.

I might be wrong but it looks like we can have only sync methods and it's the caller responsibility how it wants to handle it. 

```go
// sync
err := nodeManager.StartNode()

// async
errc := make(chan error)
go func () {
  err <- node.Manager.StartNode()
}()
// do something else in the meanwhile
err := <-errc
```

The package should not care about how it's gonna be consumed unless it's a very specific package.

Reply:
Only sync methods... I agree that it has a point. But if we have both Sync and Async methods in interface we could prevent many code duplication on client side.

Reply:
You're correct but it will cost this package:
1. More code to read and understand, thus degraded code maintainability,
2. Higher chance of maintaining unused code,
3. More unit tests,
4. Caller dilemma -- which method use: sync or async?

Also, current async implementation is very limited. Even though the caller can call an async method, it has no control over it (e.g. it can't cancel it).